zebra-chain/src/block/difficulty.rs

```diff
@@ -0,0 +1,256 @@
+//! Block difficulty data structures and calculations
+//!
+//! The block difficulty "target threshold" is stored in the block header as a
+//! 32-bit "compact bits" value. The `BlockHeaderHash` must be less than or equal
+//! to the expanded target threshold, when represented as a 256-bit integer in
+//! little-endian order.
+//!
+//! The target threshold is also used to calculate the "work" for each block.
+//! The block work is used to find the chain with the greatest total work. Each
+//! block's work value depends on the fixed threshold in the block header, not
+//! the actual work represented by the block header hash.
+
+#[cfg(test)]
+use proptest_derive::Arbitrary;
+
+/// A 32-bit "compact bits" value, which represents the difficulty threshold for
+/// a block header.
+///
+/// Used for:
+///   - checking the `difficulty_threshold` value in the block header,
+///   - calculating the 256-bit `ExpandedDifficulty` threshold, for comparison
+///     with the block header hash, and
+///   - calculating the block work.
+///
+/// Details:
+///
+/// This is a floating-point encoding, with a 24-bit signed mantissa,
+/// an 8-bit exponent, an offset of 3, and a radix of 256.
+/// (IEEE 754 32-bit floating-point values use a separate sign bit, an implicit
+/// leading mantissa bit, an offset of 127, and a radix of 2.)
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[cfg_attr(test, derive(Arbitrary))]
+pub struct CompactDifficulty(pub u32);
+
+/// A 256-bit expanded difficulty value.
+///
+/// Used as a target threshold for the difficulty of a `BlockHeaderHash`.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[cfg_attr(test, derive(Arbitrary))]
+pub struct ExpandedDifficulty([u8; 32]);
+
+impl CompactDifficulty {
+    /// CompactDifficulty floating-point precision.
+    const PRECISION: u32 = 24;
+
+    /// CompactDifficulty sign bit, part of the signed mantissa.
+    const SIGN_BIT: u32 = 1 << (CompactDifficulty::PRECISION - 1);
+
+    /// CompactDifficulty unsigned mantissa mask.
+    ///
+    /// Also the maximum unsigned mantissa value.
+    const U_MANT_MASK: u32 = CompactDifficulty::SIGN_BIT - 1;
+
+    /// CompactDifficulty exponent offset.
+    const OFFSET: i32 = 3;
+
+    /// Calculate the ExpandedDifficulty for a compact representation.
+    ///
+    /// See `ToTarget()` in the Zcash Specification, and `CheckProofOfWork()` in
+    /// zcashd.
+    ///
+    /// Returns None for negative, zero, and overflow values. (zcashd rejects
+    /// these values, before comparing the hash.)
+    pub fn to_expanded(&self) -> Option<ExpandedDifficulty> {
```

Comment:
Do we want to avoid rejecting blocks that have an invalid `ExpandedDifficulty`? If not I think we might want to do something similar to `Amount` where it uses `try_from` where we just conditionally parse straight into the ExpandedDifficulty type rather than the opaque version.

If we don't want to reject these during parsing though just completely ignore this.

Reply:
Good question!

It doesn't matter if we reject invalid difficulty values at parse time or validation time.

But we need to keep the exact `CompactDifficulty` value, because the precise bit patterns of *both* `CompactDifficulty` and `ExpandedDifficulty` are consensus-critical:
1. The `ExpandedDifficulty` value must be greater than or equal to the `BlockHeaderHash`.
2. The `CompactDifficulty` value is part of the header, so its exact bit pattern is used to create the `BlockHeaderHash`. It must also be bitwise equal to the median difficulty value from recent blocks, calculated using the specified conversion function.

Most fields have a one-to-one conversion, so the precise way we store them doesn't matter. But `CompactDifficulty` to `ExpandedDifficulty` is many-to-many:
* Due to redundancy in the `CompactDifficulty` floating-point format, there can be multiple `CompactDifficulty` values that expand to an `ExpandedDifficulty`. (If the values have 16 bits of precision or less, they can be represented as `M*256^E`, or `(M/256)*256^(E+1)`.)
* Due to rounding in the `CompactDifficulty` floating-point format, there are also multiple `ExpandedDifficulty` values that compact to each `CompactDifficulty`. (Because it only has 24 bits of precision, for a 256-bit number.)

The consensus rules handle this issue by specifying a precise compaction function, and specifying that the `CompactDifficulty` must be bitwise equal to the median recent difficulty.

I should summarise this answer as part of the module documentation.

Reply:
Documented in PR #809 in 91994501b8cb0040f765b3446f086d3393c46cd0